tokio/tests/time_interval.rs

```diff
@@ -209,3 +209,76 @@ fn poll_next(interval: &mut task::Spawn<time::Interval>) -> Poll<Instant> {
 fn ms(n: u64) -> Duration {
     Duration::from_millis(n)
 }
+
+mod tmp_tests {
+    use std::{
+        pin::Pin,
+        task::{Context, Poll},
+        time::Instant,
+    };
+
+    use crate::time::Interval;
+    use futures::{pin_mut, Stream, StreamExt};
+
+    struct IntervalStreamer {
+        start: Instant,
+        counter: u32,
+        timer: Interval,
+    }
+
+    impl Stream for IntervalStreamer {
+        type Item = u32;
+
+        fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+            let this = Pin::into_inner(self);
+
+            if this.counter > 12 {
+                return Poll::Ready(None);
+            }
+
+            match this.timer.poll_tick(cx) {
+                Poll::Pending => {
+                    println!(
+                        "Timer returned Poll::Pending after {:?}",
+                        this.start.elapsed()
+                    );
+                    Poll::Pending
+                }
+                Poll::Ready(_) => {
+                    println!(
+                        "Timer returned Poll::Ready after {:?}",
+                        this.start.elapsed()
+                    );
+                    this.counter += 1;
+                    if this.counter % 4 == 0 {
+                        Poll::Ready(Some(this.counter))
+                    } else {
+                        // Schedule this task for wake-up
+                        cx.waker().wake_by_ref();
+                        Poll::Pending
```

Comment:
As far as I can tell, you're not actually testing whether it emits a wakeup or not. This test would pass either way.

Reply:
Yes, the test was not complete but more a manual thing until I hear back on the approach. Sorry that I did not explain that better. I added two tests now which tease out the difference.